File: CleverFerret/src/main/java/com/universalmedialibrary/services/metadata/MetadataApiService.kt
package com.universalmedialibrary.services.metadata

import com.universalmedialibrary.data.repository.APIKeyRepository
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class MetadataApiService @Inject constructor(
    private val apiKeyRepository: APIKeyRepository,
    private val tvMazeService: TVMazeService
) {

    private val googleBooksApi: GoogleBooksApi by lazy {
        Retrofit.Builder()
            .baseUrl("https://www.googleapis.com/books/v1/")
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(GoogleBooksApi::class.java)
    }

    private val openLibraryApi: OpenLibraryApi by lazy {
        Retrofit.Builder()
            .baseUrl("https://openlibrary.org/")
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(OpenLibraryApi::class.java)
    }

    private val tmdbApi: TMDBApi by lazy {
        Retrofit.Builder()
            .baseUrl("https://api.themoviedb.org/3/")
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(TMDBApi::class.java)
    }

    private val musicBrainzApi: MusicBrainzApi by lazy {
        Retrofit.Builder()
            .baseUrl("https://musicbrainz.org/ws/2/")
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(MusicBrainzApi::class.java)
    }

    suspend fun searchBooks(query: String): List<UnifiedMetadataSearchResult> {
        val results = mutableListOf<UnifiedMetadataSearchResult>()

        try {
            // Google Books API
            val googleBooks = googleBooksApi.searchBooks(query)
            googleBooks.items?.forEach { item ->
                results.add(
                    UnifiedMetadataSearchResult(
                        id = item.id,
                        title = item.volumeInfo.title ?: "Unknown Title",
                        author = item.volumeInfo.authors?.joinToString(", "),
                        year = item.volumeInfo.publishedDate?.take(4)?.toIntOrNull(),
                        coverUrl = item.volumeInfo.imageLinks?.thumbnail,
                        description = item.volumeInfo.description,
                        genres = item.volumeInfo.categories ?: emptyList(),
                        rating = item.volumeInfo.averageRating,
                        isbn = item.volumeInfo.industryIdentifiers
                            ?.find { it.type == "ISBN_13" }?.identifier,
                        mediaType = com.universalmedialibrary.data.MediaType.BOOK,
                        source = MetadataSource.GOOGLE_BOOKS.name
                    )
                )
            }
        } catch (e: Exception) {
            // Log error but continue with other sources
        }

        try {
            // Open Library API
            val openLibrary = openLibraryApi.searchBooks(query)
            openLibrary.docs?.forEach { doc ->
                results.add(
                    UnifiedMetadataSearchResult(
                        id = doc.key,
                        title = doc.title ?: "Unknown Title",
                        author = doc.author_name?.joinToString(", "),
                        year = doc.first_publish_year,
                        coverUrl = doc.cover_i?.let { "https://covers.openlibrary.org/b/id/$it-M.jpg" },
                        isbn = doc.isbn?.firstOrNull(),
                        mediaType = com.universalmedialibrary.data.MediaType.BOOK,
                        source = MetadataSource.OPEN_LIBRARY.name
                    )
                )
            }
        } catch (e: Exception) {
            // Log error but continue
        }

        return results
    }

    suspend fun searchMovies(query: String): List<UnifiedMetadataSearchResult> {
        val results = mutableListOf<UnifiedMetadataSearchResult>()

        // Get TMDB API key from repository
        val tmdbApiKey = apiKeyRepository.getAPIKeyValue("tmdb")

        if (tmdbApiKey.isNullOrEmpty()) {
            // Return demo data if no API key
            return createDemoMovieResults(query)
        }

        try {
            val movies = tmdbApi.searchMovies(tmdbApiKey, query)
            movies.results?.forEach { movie ->
                results.add(
                    UnifiedMetadataSearchResult(
                        id = movie.id.toString(),
                        title = movie.title ?: movie.name ?: "Unknown Title",
                        year = movie.release_date?.take(4)?.toIntOrNull()
                            ?: movie.first_air_date?.take(4)?.toIntOrNull(),
                        coverUrl = movie.poster_path?.let { "https://image.tmdb.org/t/p/w500$it" },
                        description = movie.overview,
                        rating = movie.vote_average,
                        mediaType = com.universalmedialibrary.data.MediaType.MOVIE,
                        source = MetadataSource.TMDB.name
                    )
                )
            }
        } catch (e: Exception) {
            return createDemoMovieResults(query)
        }

        return results
    }

    suspend fun searchMusic(query: String): List<UnifiedMetadataSearchResult> {
        val results = mutableListOf<UnifiedMetadataSearchResult>()

        try {
            val music = musicBrainzApi.searchReleases(query)
            music.releases?.forEach { release ->
                results.add(
                    UnifiedMetadataSearchResult(
                        id = release.id,
                        title = release.title ?: "Unknown Title",
                        artist = release.artist_credit?.firstOrNull()?.name,
                        year = release.date?.take(4)?.toIntOrNull(),
                        mediaType = com.universalmedialibrary.data.MediaType.MUSIC,
                        source = MetadataSource.MUSICBRAINZ.name
                    )
                )
            }
        } catch (e: Exception) {
            // Return demo data on error
            return createDemoMusicResults(query)
        }

        return results
    }

    /**
     * Search for TV shows using TVMaze API
     * 
     * TVMaze provides comprehensive TV show data including:
     * - Complete episode listings
     * - Cast and crew information
     * - High-quality images
     * - Air dates and schedules
     * - Network information
     * - Links to IMDb and TheTVDB
     * 
     * Best of all: No API key required!
     * 
     * Example:
     * ```kotlin
     * val results = searchTVShows("Breaking Bad")
     * results.forEach { show ->
     *     println("${show.title} (${show.year}) - Rating: ${show.rating}")
     * }
     * ```
     */
    suspend fun searchTVShows(query: String): List<UnifiedMetadataSearchResult> {
        val results = mutableListOf<UnifiedMetadataSearchResult>()

        try {
            // Search using TVMaze API (no API key required!)
            val tvMazeResults = tvMazeService.searchShows(query)
            
            tvMazeResults.shows.forEach { show ->
                // Clean HTML from summary
                val cleanSummary = show.summary?.replace(Regex("<[^>]*>"), "")?.trim()
                
                results.add(
                    UnifiedMetadataSearchResult(
                        id = show.id.toString(),
                        title = show.name,
                        year = show.premiered?.take(4)?.toIntOrNull(),
                        coverUrl = show.image?.original ?: show.image?.medium,
                        description = cleanSummary,
                        genres = show.genres,
                        rating = show.rating?.average?.toFloat(),
                        language = show.language,
                        imdbId = show.externals?.imdb,
                        mediaType = com.universalmedialibrary.data.MediaType.TV_SHOW,
                        source = MetadataSource.TVMAZE.name,
                        additionalData = mapOf(
                            "tvMazeId" to show.id,
                            "status" to (show.status ?: "Unknown"),
                            "network" to (show.network?.name ?: show.webChannel?.name ?: "Unknown"),
                            "runtime" to (show.runtime ?: show.averageRuntime ?: 0),
                            "premiered" to (show.premiered ?: ""),
                            "ended" to (show.ended ?: ""),
                            "tvdbId" to (show.externals?.thetvdb ?: 0),
                            "officialSite" to (show.officialSite ?: "")
                        )
                    )
                )
            }
        } catch (e: Exception) {
            // Return demo data on error
            return createDemoTVShowResults(query)
        }

        return results
    }

    /**
     * Get comprehensive TV show details including episodes
     * 
     * This fetches complete show information from TVMaze including:
     * - All episodes with air dates
     * - Cast information
     * - Season information
     * - Network and streaming details
     */
    suspend fun getTVShowDetails(tvMazeId: Int): UnifiedMetadataSearchResult? {
        return try {
            val details = tvMazeService.getCompleteShowDetails(tvMazeId)
            val show = details.show ?: return null
            
            val cleanSummary = show.summary?.replace(Regex("<[^>]*>"), "")?.trim()
            
            UnifiedMetadataSearchResult(
                id = show.id.toString(),
                title = show.name,
                year = show.premiered?.take(4)?.toIntOrNull(),
                coverUrl = show.image?.original ?: show.image?.medium,
                description = cleanSummary,
                genres = show.genres,
                rating = show.rating?.average?.toFloat(),
                language = show.language,
                imdbId = show.externals?.imdb,
                mediaType = com.universalmedialibrary.data.MediaType.TV_SHOW,
                source = MetadataSource.TVMAZE.name,
                additionalData = mapOf(
                    "tvMazeId" to show.id,
                    "status" to (show.status ?: "Unknown"),
                    "network" to (show.network?.name ?: show.webChannel?.name ?: "Unknown"),
                    "runtime" to (show.runtime ?: show.averageRuntime ?: 0),
                    "premiered" to (show.premiered ?: ""),
                    "ended" to (show.ended ?: ""),
                    "tvdbId" to (show.externals?.thetvdb ?: 0),
                    "officialSite" to (show.officialSite ?: ""),
                    "totalEpisodes" to details.totalEpisodes,
                    "totalSeasons" to details.totalSeasons,
                    "episodes" to details.episodes,
                    "cast" to details.cast,
                    "seasons" to details.seasons
                )
            )
        } catch (e: Exception) {
            null
        }
    }

    private fun createDemoMovieResults(query: String): List<UnifiedMetadataSearchResult> {
        return listOf(
            UnifiedMetadataSearchResult(
                id = "demo1",
                title = "Demo Movie: $query",
                director = "Demo Director",
                year = 2024,
                coverUrl = "https://via.placeholder.com/500x750/1565C0/ffffff?text=Movie",
                description = "This is a demo movie result for '$query'. In production, this would use the actual TMDB API.",
                genres = listOf("Action", "Drama"),
                rating = 8.5f,
                mediaType = com.universalmedialibrary.data.MediaType.MOVIE,
                source = MetadataSource.TMDB.name
            )
        )
    }

    private fun createDemoMusicResults(query: String): List<UnifiedMetadataSearchResult> {
        return listOf(
            UnifiedMetadataSearchResult(
                id = "demo1",
                title = "Demo Album: $query",
                artist = "Demo Artist",
                year = 2024,
                coverUrl = "https://via.placeholder.com/500x500/7B1FA2/ffffff?text=Music",
                description = "This is a demo music result for '$query'.",
                genres = listOf("Electronic", "Ambient"),
                mediaType = com.universalmedialibrary.data.MediaType.MUSIC,
                source = MetadataSource.MUSICBRAINZ.name
            )
        )
    }

    private fun createDemoTVShowResults(query: String): List<UnifiedMetadataSearchResult> {
        return listOf(
            UnifiedMetadataSearchResult(
                id = "demo1",
                title = "Demo TV Show: $query",
                director = "Demo Creator",
                year = 2024,
                coverUrl = "https://via.placeholder.com/500x750/FF9800/ffffff?text=TV+Show",
                description = "This is a demo TV show result for '$query'. In production, this would use the TVMaze API.",
                genres = listOf("Drama", "Sci-Fi"),
                rating = 8.5f,
                mediaType = com.universalmedialibrary.data.MediaType.TV_SHOW,
                source = MetadataSource.TVMAZE.name
            )
        )
    }
}

File: pwa-demo/src/services/metadata/MetadataApiService.ts
/**
 * Metadata API Service
 * 
 * Fetches metadata from various external APIs (Google Books, TMDB, etc.)
 * Migrated from MetadataApiService.kt and RealMetadataService.kt
 */

import { networkManager } from '../network/NetworkManager';
import { API_ENDPOINTS } from '../../index';

export class MetadataApiService {
  /**
   * Search book metadata from Google Books API
   */
  async searchBookMetadata(
    title: string,
    author?: string,
    isbn?: string
  ): Promise<BookMetadataResult> {
    try {
      let query = '';
      if (isbn) {
        query = `isbn:${encodeURIComponent(isbn)}`;
      } else {
        query = `intitle:${encodeURIComponent(title)}`;
        if (author) {
          query += `+inauthor:${encodeURIComponent(author)}`;
        }
      }

      const response = await networkManager.get<GoogleBooksResponse>(
        `${API_ENDPOINTS.GOOGLE_BOOKS}/volumes?q=${query}&maxResults=5`
      );

      if (!response.success || !response.data?.items?.length) {
        return {
          success: false,
          error: 'No results found',
        };
      }

      const book = response.data.items[0].volumeInfo;

      return {
        success: true,
        metadata: {
          title: book.title,
          authors: book.authors || [],
          publisher: book.publisher,
          publishedDate: book.publishedDate,
          description: book.description,
          isbn: book.industryIdentifiers?.find(id => id.type === 'ISBN_13')?.identifier,
          pageCount: book.pageCount,
          coverUrl: book.imageLinks?.thumbnail || book.imageLinks?.smallThumbnail,
          language: book.language,
          categories: book.categories || [],
          averageRating: book.averageRating,
          ratingsCount: book.ratingsCount,
          sourceId: response.data.items[0].id,
        },
        sources: ['Google Books'],
      };
    } catch (error: any) {
      return {
        success: false,
        error: error.message || 'Failed to fetch book metadata',
      };
    }
  }

  /**
   * Search movie metadata from TMDB API
   */
  async searchMovieMetadata(
    title: string,
    year?: number,
    imdbId?: string
  ): Promise<MovieMetadataResult> {
    try {
      // Note: TMDB requires an API key - this is a placeholder
      const apiKey = ''; // User would need to provide their own key

      if (!apiKey) {
        return {
          success: false,
          error: 'TMDB API key not configured',
        };
      }

      let endpoint = '';
      if (imdbId) {
        endpoint = `${API_ENDPOINTS.TMDB}/find/${imdbId}?api_key=${apiKey}&external_source=imdb_id`;
      } else {
        let query = `${API_ENDPOINTS.TMDB}/search/movie?api_key=${apiKey}&query=${encodeURIComponent(title)}`;
        if (year) {
          query += `&year=${year}`;
        }
        endpoint = query;
      }

      const response = await networkManager.get<TMDBSearchResponse>(endpoint);

      if (!response.success || !response.data?.results?.length) {
        return {
          success: false,
          error: 'No results found',
        };
      }

      const movie = response.data.results[0];

      return {
        success: true,
        metadata: {
          title: movie.title,
          originalTitle: movie.original_title,
          overview: movie.overview,
          releaseDate: movie.release_date,
          year: movie.release_date ? parseInt(movie.release_date.substring(0, 4)) : undefined,
          posterUrl: movie.poster_path ? `https://image.tmdb.org/t/p/w500${movie.poster_path}` : undefined,
          backdropUrl: movie.backdrop_path ? `https://image.tmdb.org/t/p/w1280${movie.backdrop_path}` : undefined,
          voteAverage: movie.vote_average,
          voteCount: movie.vote_count,
          popularity: movie.popularity,
          tmdbId: movie.id?.toString(),
          imdbId: imdbId || undefined,
        },
        sources: ['TMDB'],
      };
    } catch (error: any) {
      return {
        success: false,
        error: error.message || 'Failed to fetch movie metadata',
      };
    }
  }

  /**
   * Search music metadata from MusicBrainz API
   */
  async searchMusicMetadata(
    artist?: string,
    album?: string,
    track?: string
  ): Promise<MusicMetadataResult> {
    try {
      const parts: string[] = [];
      if (track) parts.push(`recording:"${track}"`);
      if (artist) parts.push(`artist:"${artist}"`);
      if (album) parts.push(`release:"${album}"`);
      const query = parts.join(' AND ');

      const response = await networkManager.get<MusicBrainzResponse>(
        `${API_ENDPOINTS.MUSICBRAINZ}/recording?query=${encodeURIComponent(query)}&fmt=json&limit=5`
      );

      if (!response.success || !response.data?.recordings?.length) {
        return {
          success: false,
          error: 'No results found',
        };
      }

      const recording = response.data.recordings[0];

      return {
        success: true,
        metadata: {
          title: recording.title,
          artist: recording['artist-credit']?.[0]?.name,
          album: recording.releases?.[0]?.title,
          duration: recording.length,
          releaseDate: recording.releases?.[0]?.date,
          mbid: recording.id,
          coverUrl: undefined, // MusicBrainz doesn't provide cover art directly
        },
        sources: ['MusicBrainz'],
      };
    } catch (error: any) {
      return {
        success: false,
        error: error.message || 'Failed to fetch music metadata',
      };
    }
  }

  /**
   * Get cover art from Cover Art Archive
   */
  async getCoverArt(mbid: string): Promise<string | null> {
    try {
      const response = await networkManager.get<any>(
        `https://coverartarchive.org/release/${mbid}`
      );

      if (response.success && response.data?.images?.length) {
        return response.data.images[0].image;
      }

      return null;
    } catch (error) {
      console.warn(`Failed to fetch cover art for MBID: ${mbid}`, error);
      return null;
    }
  }
}

// Type definitions

interface GoogleBooksResponse {
  items?: Array<{
    id: string;
    volumeInfo: {
      title: string;
      authors?: string[];
      publisher?: string;
      publishedDate?: string;
      description?: string;
      industryIdentifiers?: Array<{ type: string; identifier: string }>;
      pageCount?: number;
      imageLinks?: {
        thumbnail?: string;
        smallThumbnail?: string;
      };
      language?: string;
      categories?: string[];
      averageRating?: number;
      ratingsCount?: number;
    };
  }>;
}

interface TMDBSearchResponse {
  results?: Array<{
    id: number;
    title: string;
    original_title: string;
    overview: string;
    release_date: string;
    poster_path?: string;
    backdrop_path?: string;
    vote_average: number;
    vote_count: number;
    popularity: number;
  }>;
}

interface MusicBrainzResponse {
  recordings?: Array<{
    id: string;
    title: string;
    length?: number;
    'artist-credit'?: Array<{ name: string }>;
    releases?: Array<{
      title: string;
      date?: string;
    }>;
  }>;
}

export interface BookMetadata {
  title: string;
  authors: string[];
  publisher?: string;
  publishedDate?: string;
  description?: string;
  isbn?: string;
  pageCount?: number;
  coverUrl?: string;
  language?: string;
  categories: string[];
  averageRating?: number;
  ratingsCount?: number;
  sourceId?: string;
}

export interface MovieMetadata {
  title: string;
  originalTitle?: string;
  overview?: string;
  releaseDate?: string;
  year?: number;
  posterUrl?: string;
  backdropUrl?: string;
  voteAverage?: number;
  voteCount?: number;
  popularity?: number;
  tmdbId?: string;
  imdbId?: string;
}

export interface MusicMetadata {
  title: string;
  artist?: string;
  album?: string;
  duration?: number;
  releaseDate?: string;
  mbid?: string;
  coverUrl?: string;
}

export interface BookMetadataResult {
  success: boolean;
  metadata?: BookMetadata;
  sources?: string[];
  error?: string;
}

export interface MovieMetadataResult {
  success: boolean;
  metadata?: MovieMetadata;
  sources?: string[];
  error?: string;
}

export interface MusicMetadataResult {
  success: boolean;
  metadata?: MusicMetadata;
  sources?: string[];
  error?: string;
}

// Export singleton instance
export const metadataApiService = new MetadataApiService();
